### PR TITLE
Remove code using `spark.client._builder` in new version DBconnect

### PR DESCRIPTION
**Pull request by WeichenXu123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/WeichenXu123/mlflow/pull/13549?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13549/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13549
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Remove code using `spark.client._builder` in new version DBconnect

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [X] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/utils/databricks_utils.py [hunk 1]
```diff
@@ -288,16 +288,13 @@ def is_databricks_connect(spark=None):
...
-        if spark is None:
-            spark = _get_active_spark_session()
-        # TODO: Remove the `spark.client._builder` attribute usage once
-        #  Spark-connect has public attribute for this information.
-        return is_spark_connect_mode() and any(
-            k == "x-databricks-cluster-id" for k, v in spark.client._builder.metadata()
-        )
-    except Exception:
-        return False
+
```
Comment (serena-ruan) [inline mlflow/utils/databricks_utils.py:291]: I updated `spark=None` so we still need 
```
if spark is None:
            spark = _get_active_spark_session()
```

File: mlflow/utils/databricks_utils.py [hunk 2]
```diff
@@ -288,16 +288,13 @@ def is_databricks_connect(spark=None):
...
+    if hasattr(spark.client, "metadata"):
```
Comment (serena-ruan) [inline mlflow/utils/databricks_utils.py:292]: Do we need check `hasattr(spark, "client")`? It's possible `spark` is None

File: mlflow/utils/databricks_utils.py [hunk 3]
```diff
@@ -358,16 +355,16 @@ def is_databricks_serverless(spark):
...
-    except Exception:
+    if not is_spark_connect_mode():
         return False
 
+    if hasattr(spark.client, "metadata"):
+        metadata = spark.client.metadata
+    else:
+        metadata = spark.client._builder.metadata()
+
+    return any(k == "x-databricks-session-id" for k, v in metadata)
```
Comment (serena-ruan) [inline mlflow/utils/databricks_utils.py:366]: what type is metadata? list[tuple]?

Comment (WeichenXu123) [reply mlflow/utils/databricks_utils.py:366]: yes

File: mlflow/utils/databricks_utils.py (hunk 2 above)
Comment (WeichenXu123) [reply mlflow/utils/databricks_utils.py:295]: no.

now code is like
```
    if spark is None:
        spark = _get_active_spark_session()

    if not is_spark_connect_mode():
        return False

    if hasattr(spark.client, "metadata"):
```

if it is sparkconnect, then spark must have client attribute

Comment (serena-ruan) [reply mlflow/utils/databricks_utils.py:295]: Maybe check if spark is None, if so return False :)

Comment (serena-ruan) [reply mlflow/utils/databricks_utils.py:295]:
```
def is_spark_connect_mode():
    try:
        from pyspark.sql.utils import is_remote
    except ImportError:
        return False
    return is_remote()

def is_remote() -> bool:
    """
    Returns if the current running environment is for Spark Connect.
    """
    return "SPARK_CONNECT_MODE_ENABLED" in os.environ
```
This doesn't mean spark is not None right?
Currently `if not is_spark_connect_mode(): return False` check is at the top

File: mlflow/utils/databricks_utils.py [hunk 4]
```diff
@@ -283,22 +283,22 @@ def is_databricks_connect(spark=None):
...
-            spark = _get_active_spark_session()
-        # TODO: Remove the `spark.client._builder` attribute usage once
-        #  Spark-connect has public attribute for this information.
-        return is_spark_connect_mode() and any(
-            k == "x-databricks-cluster-id" for k, v in spark.client._builder.metadata()
-        )
-    except Exception:
+
+    if spark is None:
+        spark = _get_active_spark_session()
```
Comment (serena-ruan) [inline mlflow/utils/databricks_utils.py:290]:
```suggestion
    spark = spark or _get_active_spark_session()
    if spark is None:
        return False
```

**Review by serena-ruan — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/13549/files#r1827250901 is addressed!

File: mlflow/utils/databricks_utils.py (hunk 2 above)
Comment (WeichenXu123) [reply mlflow/utils/databricks_utils.py:296]: oh :) you are right :)

**Pull request merged**